### PR TITLE
[mcs] extend API for loading types and consuming source file from a stream

### PR DESCRIPTION
**Pull request by lewurm**
contributed by @SeargeDP

### CODE REVIEW

**Comment by lewurm** [comment]
@marek-safar is there a way to run the `mcs` test-suite despite `--with-csc=roslyn`?

File: mcs/mcs/location.cs [hunk 1]
```diff
@@ -53,22 +53,26 @@ public override string ToString ()
...
 
+		public delegate SeekableStreamReader GetInputStreamDelegate(SourceFile sourceFile);
+
 		static readonly byte[] MD5Algorith = { 96, 166, 110, 64, 207, 100, 130, 76, 182, 240, 66, 212, 129, 114, 167, 153 };
 
 		public readonly string Name;
 		public readonly string OriginalFullPathName;
 		public readonly int Index;
 		public bool AutoGenerated;
+		public GetInputStreamDelegate GetInputStream;
```
Comment (marek-safar) [inline mcs/mcs/location.cs]: Use Func instead of custom delegate

File: mcs/mcs/location.cs [hunk 2]
```diff
@@ -53,22 +53,26 @@ public override string ToString ()
...
 
 		SourceFileEntry file;
 		byte[] algGuid, checksum;
 		List<LocationRegion> hidden_lines;
 
-		public SourceFile (string name, string path, int index)
+		public SourceFile (string name, string path, int index, GetInputStreamDelegate inputStreamDelegate)
```
Comment (marek-safar) [inline mcs/mcs/location.cs]: Add another ctor overload with GetInputStreamDelegate

File: mcs/mcs/eval.cs [hunk 3]
```diff
@@ -958,7 +958,14 @@ public void ReferenceAssembly (Assembly a)
 				importer.ImportAssembly (a, module.GlobalRootNamespace);
 			}
 		}
-	}
+
+		public void ImportTypes(bool importExtensionTypes, params Type[] types) {
```
Comment (marek-safar) [inline mcs/mcs/eval.cs]: Please follow existing code style here and everywhere, especially space before `(`.
This block looks off

File: mcs/mcs/driver.cs [hunk 4]
```diff
@@ -129,36 +143,50 @@ void ParseParallel (ModuleContainer module)
...
 
-			Parse (reader, file, module, session, report);
+					DoParse(file, module, session, report, reader);
+				}
+			} else {
+				DoParse(file, module, session, report, reader);
+			}
+		}
+
+		private void DoParse(SourceFile file, ModuleContainer module, ParserSession session, Report report, SeekableStreamReader reader) {
```
Comment (marek-safar) [inline mcs/mcs/driver.cs]: redundant private

**Comment by marek-safar** [comment]
@lewurm they are ran in part of any PR or you can go to mcs/tests and make check then go to mcs/class/Mono.CSharp do same

**Comment by lewurm** [comment]
`make -C tests check && make -C class/Mono.CSharp check` passes successfully on my machine now, thanks for the feedback @marek-safar

File: mcs/mcs/location.cs [hunk 5]
```diff
@@ -59,16 +59,22 @@ public override string ToString ()
...
 		SourceFileEntry file;
 		byte[] algGuid, checksum;
 		List<LocationRegion> hidden_lines;
 
-		public SourceFile (string name, string path, int index)
+		public SourceFile (string name, string path, int index) : this (name, path, index, null)
+		{
+		}
+
+		public SourceFile (string name, string path, int index, Func<SourceFile, SeekableStreamReader> inputStreamDelegate)
```
Comment (marek-safar) [inline mcs/mcs/location.cs]: Please do it the other way around, call this from this overload

File: mcs/mcs/driver.cs [hunk 6]
```diff
@@ -44,31 +44,45 @@ public Driver (CompilerContext ctx)
...
+				using (input) {
+					reader = new SeekableStreamReader (input, ctx.Settings.Encoding);
+					DoTokenize (sourceFile, module, session, reader);
+				}
+			} else {
+				DoTokenize (sourceFile, module, session, reader);
+			}
+		}
+
+		private void DoTokenize (SourceFile sourceFile, ModuleContainer module, ParserSession session, SeekableStreamReader reader) {
```
Comment (marek-safar) [inline mcs/mcs/driver.cs]: One more redundant private

**Comment by lewurm** [comment]
thanks @marek-safar. PR updated

**Pull request merged**